pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -290,7 +290,11 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
                 "Disconnected from server at " + ctx.channel().remoteAddress());
 
         // Fail out all the pending ops
-        pendingRequests.forEach((key, future) -> future.completeExceptionally(e));
+        pendingRequests.forEach((key, future) -> {
+            if (pendingRequests.remove(key, future) && !future.isDone()) {
```

Comment:
is it really allowed and officially supported to mutate the collection inside a forEach loop ?
are there any side effects ?

Reply:
Great suggestion, I'll check if it works.

Reply:
I just tested it. Elements can be deleted as expected in `for each`. And because when the channel is Inactive, this ClientCnx will not be reused but recycled by the JVM, so we don't really care whether it is really deleted. Therefore, there will be no related side effects.